### PR TITLE
Add `cupy.typing`

### LINKED ISSUE #6244: Add `numpy.typing` equivalent APIs in CuPy
This will allow adding type annotations to CuPy's codebase (#4831) and users' code.

https://numpy.org/devdocs/reference/typing.html

### PR DESCRIPTION
**Pull request by asi1024**
Close #6244.

### CODE REVIEW

File: cupy/typing/_alias.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from typing import Any, TypeVar
+
+import numpy
+import numpy.typing as npt
+
+import cupy
+
+
+if numpy.lib.NumpyVersion(numpy.__version__) < '1.21.0':
+    raise ImportError('`cupy.typing` requires NumPy 1.21 or later.')
```
Comment (toslunar) [inline cupy/typing/_alias.py:10]: Does it cost to support NumPy 1.20?

File: cupy/typing/_alias.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+
+ArrayLike = npt.ArrayLike
+DTypeLike = npt.DTypeLike
+NBitBase = npt.NBitBase
+ScalarType = TypeVar("ScalarType", bound=numpy.generic, covariant=True)
+_DType = npt._GenericAlias(numpy.dtype, (ScalarType,))
+NDArray = npt._GenericAlias(cupy.ndarray, (Any, _DType))
```
Comment (toslunar) [inline cupy/typing/_alias.py:18]: `if TYPE_CHECKING or sys.version_info >= (3, 9)`, NumPy uses class-getitem `NDArray = np.ndarray[Any, np.dtype[ScalarType]]`.
numpy/numpy#19879

File: cupy/typing/_alias.py (hunk 1 above)
Comment (asi1024) [reply cupy/typing/_alias.py]: If we support numpy<1.21 and python<3.9, we need to reimplement `GenericAlias` in CuPy's codebase.

**Review by toslunar — APPROVED** [review]
LGTM

**Comment by kmaehashi** [comment]
Do we have to make `GenericAlias` available for all supported NumPy/Python versions, otherwise we can't use `NDArray` within CuPy codebase?

**Comment by kmaehashi** [comment]
Let's borrow it from a tagged NumPy source tree 😃 
https://github.com/numpy/numpy/blob/v1.22.1/numpy/typing/_generic_alias.py

File: cupy/typing/_generic_alias.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    from types import GenericAlias
+elif numpy.lib.NumpyVersion(numpy.__version__) >= '1.21.0':
+    from numpy.typing import _GenericAlias as GenericAlias
+else:
+    def GenericAlias(*args):
+        return Any
+
+
+_ScalarType = TypeVar("ScalarType", bound=numpy.generic, covariant=True)
+_DType = GenericAlias(numpy.dtype, (_ScalarType))
```
Comment (toslunar) [inline cupy/typing/_generic_alias.py:29]:
```suggestion
_DType = GenericAlias(numpy.dtype, (_ScalarType,))
```

**Review by toslunar — APPROVED** [review]

**Pull request merged**